### PR TITLE
Updates to `snarkvm-parameters` and expands serial numbers tree and commitments tree to depth 42

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

- Upgrades the `snarkvm-parameters` setup procedure to be more streamlined and less error-prone.
- Expands serial numbers tree and commitments tree to depth `42`
- Unifies the given owner allocation for input records in the inner circuit
- Removes `TransactionScheme` trait

### CODE REVIEW

File: gadgets/src/bits/boolean_input.rs [hunk 1]
```diff
@@ -95,20 +95,49 @@ impl<F: PrimeField, CF: PrimeField> AllocGadget<Vec<F>, CF> for BooleanInputGadg
...
-        let mut res = Vec::<Vec<Boolean>>::new();
-        for (i, elem) in obj.borrow().iter().enumerate() {
+        // Step 1: obtain the bits of the F field elements
+        let mut src_bits = Vec::<bool>::new();
+        for elem in obj.borrow().iter() {
             let mut bits = elem.to_repr().to_bits_le();
             bits.truncate(F::size_in_bits());
+            for _ in bits.len()..F::size_in_bits() {
+                bits.push(false);
+            }
```
Comment (ljedrz) [inline gadgets/src/bits/boolean_input.rs:105]:
```suggestion
            bits.resize(F::size_in_bits(), false);
```
(untested, but should work)

File: parameters/examples/setup.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        Some(sum) => format!("{}.{}", filename, sum),
+        _ => filename.to_string(),
+    }
+}
+
+/// Writes the given bytes to the given versioned filename.
+fn write_remote(filename: &str, version: &str, bytes: &[u8]) -> Result<()> {
+    let mut file = BufWriter::new(File::create(PathBuf::from(&versioned_filename(filename, version)))?);
+    file.write_all(&bytes)?;
+    drop(file);
```
Comment (ljedrz) [inline parameters/examples/setup.rs:54]: `file` will get dropped automatically at the end of the function (applies to the other functions here as well)

File: parameters/examples/setup.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    Ok(())
+}
+
+/// Writes the given bytes to the given filename.
+fn write_local(filename: &str, bytes: &[u8]) -> Result<()> {
+    let mut file = BufWriter::new(File::create(PathBuf::from(filename))?);
+    file.write_all(&bytes)?;
+    drop(file);
```
Comment (ljedrz) [inline parameters/examples/setup.rs:62]:
```suggestion
```

File: parameters/examples/setup.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    Ok(())
+}
+
+/// Writes the given metadata as JSON to the given filename.
+fn write_metadata(filename: &str, metadata: &Value) -> Result<()> {
+    let mut file = BufWriter::new(File::create(PathBuf::from(filename))?);
+    file.write_all(&serde_json::to_vec_pretty(metadata)?)?;
+    drop(file);
```
Comment (ljedrz) [inline parameters/examples/setup.rs:70]:
```suggestion
```

**Comment by Pratyush** [comment]
Why depth 42? What does that buy us? I imagine in a future update we'll have to switch serial number formats much before we hit 2^32 sns, no?

**Review by raychu86 — APPROVED** [review]

**Review by collinc97 — APPROVED** [review]

**Comment by howardwu** [comment]
> Why depth 42? What does that buy us? I imagine in a future update we'll have to switch serial number formats much before we hit 2^32 sns, no?

Great question, the rationale of 2^32 is to that 2^32 would require us to flush the Merkle tree once per year (assuming ~1M blocks per year with 500 txs on average per block). Going to 2^42 for 500 txs per block matches perfectly to 2^32 block height (which is what block heights are encoded as at the moment)

**Comment by howardwu** [comment]
As others are waiting for this PR to land, I will preemptively merge it as it is in stable state, and continue finishing the feature in a new PR

**Pull request merged**